R/run-nmtran.R

```diff
@@ -0,0 +1,223 @@
+#' generate `NM-TRAN` datasets (`FDATA`).
+#'
+#' @details
+#' `NM-TRAN` is a preprocessor for `NONMEM` that translates user-specified
+#' control stream data and instructions into a form executable by `NONMEM`.
+#'
+#' `run_nmtran()` allows users to test their models ahead of submission to ensure
+#' correct coding.
+#'
+#' @section Supported `bbi_args`:
+#' `run_nmtran()` supports passing the following raw `NMFE` options through to `NM-TRAN`
+#' via the familiar `.bbi_args` argument.
+#'  - **`maxlim`**: Set the maximum values for the buffers used by `NONMEM`
+#'  (if 0, don't pass `-maxlim` to nmfe) (default is 2)
+#'    - If `maxlim = 3`, it is preferred to also set `tprdefault = TRUE`, but
+#'    _not_ `prdefault`, as `NM-TRAN`'s optional resizing of the `PREDPP` size
+#'    parameter `MAXRECID` may conflict with the `-prdefault` option.
+#'  - **`prdefault`**: If `TRUE`, do not recompile any routines other than `FSUBS`
+#'  - **`tprdefault`**: If `TRUE`, test if is okay to do `-prdefault`
+#'
+#' @param .mod A `bbr` model object.
+#' @param .bbi_args A named list specifying arguments to pass to `NM-TRAN`.
+#'   Similar to the `.bbi_args` argument defined in [submit_model()], though here
+#'   only `prdefault`, `tprdefault`, and `maxlim` flags are passed to `NM-TRAN`.
+#'   See details.
+#' @param .config_path Path to a bbi configuration file. If `NULL`, the default,
+#'   will attempt to use a `bbi.yaml` in the same directory as the model.
+#' @param nmtran_exe Path to an `NM-TRAN` executable. If `NULL`, will look for a
+#'   `bbi.yaml` file in the same directory as the model.
+#' @param delete_on_exit Logical. If `FALSE`, don't delete the temporary folder
+#'   containing the `NM-TRAN` run, which will be stored in the current working
+#'   directory.
+#'
+#' @examples
+#' \dontrun{
+#' mod <- read_model(file.path(MODEL_DIR, 1))
+#' run_nmtran(mod)
+#'
+#' # Set the path to an NM-TRAN executable
+#' run_nmtran(mod, nmtran_exe = "/opt/NONMEM/nm75/tr/NMTRAN.exe")
+#'
+#' }
+#' @name nmtran
+NULL
+
+
+#' @describeIn nmtran Run `NM-TRAN` on a model object to validate its control
+#' stream for correct coding before submission.
+#' @export
+run_nmtran <- function(
+    .mod,
+    .bbi_args = list(prdefault = FALSE, tprdefault = FALSE, maxlim = 2),
+    .config_path = NULL,
+    nmtran_exe = NULL,
+    delete_on_exit = TRUE
+){
+  check_model_object(.mod, "bbi_nonmem_model")
+
+  # Capture NONMEM and NMTRAN options
+  nmtran_exe <- locate_nmtran(.mod, .config_path, nmtran_exe)
+  nm_ver <- attr(nmtran_exe, "nonmem_version")
+
+  # Combine NONMEM submission args
+  #  - The main ones of interest are prdefault, tprdefault, and maxlim, which
+  # impact the evaluation of `NM-TRAN`
+  .bbi_args <- parse_args_list(.bbi_args, .mod[[YAML_BBI_ARGS]])
+  .bbi_args <- Filter(Negate(is.null), .bbi_args[c("prdefault", "tprdefault", "maxlim")])
+  cmd_args <- check_bbi_args(.bbi_args)
```

Comment:
nmfe75 calls NMTRAN with three or four positional args. nmfe73 and nmfe74 call it with three. As mentioned before, I think we're fine ignoring 75's four-arg variant.

-   <https://github.com/metrumresearchgroup/bbr/issues/650#issuecomment-1930725642>
-   <https://github.com/metrumresearchgroup/bbr/pull/705#discussion_r1677878012>

So the goal is to call NMTRAN in the same way, passing three positional args: prdefault, tprdefault, and maxlim. The logic here fails to do that. By default, it passes a one-item vector, "&#x2013;maxlim=2", as the only argument to NMTRAN. That's wrong for two reasons:

-   it's getting passed in the first position, so NMTRAN would take that as the prdefault value.

-   it's passed with the option &#x2013;maxlim tacked on, but that's not how NMTRAN receives the value. NMTRAN receives just "2".

NMTRAN should always be getting called with three positional arguments. We need to map these onto what NMTRAN would see when we invoke bbi/nmfe with a given option. See the discussion in the [my previous comment](https://github.com/metrumresearchgroup/bbr/pull/705#discussion_r1677878012) about "matching what would be triggered by a vanilla bbi".

An additional complication here is that this doesn't respect arguments specified in `bbi.yaml`, like bbi would. I think that's probably okay, though it should definitely be documented.

It's various complications like these that make it tempting to [just document the hard-coded arguments](https://github.com/metrumresearchgroup/bbr/pull/705#discussion_r1680014896) we pass (picked to match what happens with vanilla bbi).

Reply:
I think I resolved this, though I now see a different default output (see below). Not sure if the order is wrong (I know you said you had to confirm it), but will look into it after I finish addressing some of the other feedback. 

New output:
```r
 WARNINGS AND ERRORS (IF ANY) FOR PROBLEM    1
             
 (WARNING  2) NM-TRAN INFERS THAT THE DATA ARE POPULATION.

 LIM VALUES MAXLIM ASSESSED BY NMTRAN: 1,2,3,4,5,6,7,8,10,11,13,15,16        
  
Note: Analytical 2nd Derivatives are constructed in FSUBS but are never used.
      You may insert $ABBR DERIV2=NO after the first $PROB to save FSUBS construction and compilation time
```

Reply:
> Not sure if the order is wrong (I know you said you had to confirm it)

That comment was about the values under a vanilla bbi run, not the order.  The order is known by looking at the NMTRAN.exe calls in the nmfe* files.

Here's confirmation that it's `"0" "0" "2"`:

<details>
<summary>script</summary>

```sh
#!/bin/sh

set -eu

tdir=$(mktemp -d "${TMPDIR:-/tmp}"/bbi-nmtran-XXXXXXX)
cd "$tdir"

url_base=https://raw.githubusercontent.com/metrumresearchgroup/bbi/v3.3.0/integration/testdata/acop
curl -fSsL "$url_base/acop.mod" >1.mod
curl -fSsLO "$url_base/acop.csv"

bbi version
bbi init --dir /opt/NONMEM

cat <<'EOF' >nmfe75.patch
--- nmfe75	2024-07-23 19:42:21.460325329 -0400
+++ nmfe75	2024-07-23 19:44:07.685054604 -0400
@@ -189,6 +189,8 @@
    cp tempzzzz1 "$1"_dde
    fi
 
+  printf 'NMTRAN args\n  prdefault: "%s"\n  tprdefault: "%s"\n  maxlim: "%s"\n  do2test: "%s"' \
+         "$prdefault" "$tprdefault" "$maxlim" "$do2test"
   $dir/util/nmtran_presort < tempzzzz1 | $dir/tr/NMTRAN.exe $prdefault $tprdefault $maxlim $do2test >& FMSG
   status=$?
   if [  $status -eq 9 ]
EOF

sha256sum /opt/NONMEM/nm75/run/nmfe75
sudo patch -d /opt/NONMEM/nm75/run/ <nmfe75.patch

bbi nonmem run local --clean_lvl=0 --nm_version=nm75 1.mod

sudo patch -R -d /opt/NONMEM/nm75/run/ <nmfe75.patch
sha256sum /opt/NONMEM/nm75/run/nmfe75

grep -A4 'NMTRAN args' 1/1.mod.out
```

</details>

```
v3.3.0
74b3b663e3723fa9ec84cbaf2d96b2c2d4b201e6c1b3c1ca9b52fa0d20197d1d  /opt/NONMEM/nm75/run/nmfe75
patching file nmfe75
INFO[0000] Successfully loaded default configuration from /tmp/bbi-nmtran-78o0Q2s/bbi.yaml 
INFO[0000] Beginning Local Path                         
INFO[0000] A total of 1 models have completed the initial preparation phase 
INFO[0000] [1] Beginning local work phase               
INFO[0007] [1] Beginning cleanup phase                  
INFO[0007] [1] Cleanup completed                        
INFO[0007] Waiting for any post execution hooks to finish 
1 models completed in 7.650812842s          

patching file nmfe75
74b3b663e3723fa9ec84cbaf2d96b2c2d4b201e6c1b3c1ca9b52fa0d20197d1d  /opt/NONMEM/nm75/run/nmfe75
NMTRAN args
  prdefault: "0"
  tprdefault: "0"
  maxlim: "2"
  do2test: "0"
```

---

Side remark: under that execution, nmfe75's first call to NMTRAN (with `do2test`) doesn't exit with the exit code (9) that sends it down the path that makes a second NMTRAN call without `do2test`, so it's worth thinking a bit more about whether _not_ passing do2test for nmfe75 is okay.  That is, revisit my thought [here](https://github.com/metrumresearchgroup/bbr/pull/705#discussion_r1677878012) that "it's probably safe to just ignore this upfront call given bbi doesn't expose the argument".

A first step to that might be check whether NMTRAN invocations without a fourth argument and with "0" as the fourth argument seem to produce the same output.  It'd also be helpful to find documentation on what `do2test` is about, but I haven't had any luck.